src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+        };
+
+        let bb1 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: bb1_ext.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&bb1_ext),
+            make_opaque(
+                PRELUDE_ID,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&bb1_ext),
+            ),
+        )?;
+
+        let bb2 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: bb2_ext.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bb2_ext),
+            make_opaque(
+                PRELUDE_ID,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bb2_ext),
+            ),
+        )?;
+
+        hugr.connect(entry, 0, bb1, 0)?;
+        hugr.connect(bb1, 0, bb2, 0)?;
+        hugr.connect(bb1, 0, exit, 0)?;
+        hugr.connect(bb2, 0, entry, 0)?;
+
+        Ok(hugr)
+    }
+
+    #[test]
+    fn test_cfg_loops() -> Result<(), Box<dyn Error>> {
+        let just_a = ExtensionSet::singleton(&A);
+        let variants = vec![
+            (
+                ExtensionSet::new(),
+                ExtensionSet::new(),
+                ExtensionSet::new(),
+            ),
+            (just_a.clone(), ExtensionSet::new(), ExtensionSet::new()),
+            (ExtensionSet::new(), just_a.clone(), ExtensionSet::new()),
+            (ExtensionSet::new(), ExtensionSet::new(), just_a.clone()),
+        ];
+
+        for (bb0, bb1, bb2) in variants.into_iter() {
+            println!("{}, {}, {}", bb0, bb1, bb2);
+            let mut hugr = make_cfg_loop_test(bb0, bb1, bb2)?;
+            hugr.infer_and_validate(&PRELUDE_REGISTRY)?;
+        }
+        Ok(())
+    }
+
+    #[test]
+    /// A control flow graph consisting of an entry node and a single block
+    /// which adds a resource and links to both itself and the exit node.
+    fn simple_cfg_loop() -> Result<(), Box<dyn Error>> {
+        let just_a = ExtensionSet::singleton(&A);
+
+        let mut hugr = Hugr::new(NodeType::new(
+            ops::CFG {
+                signature: FunctionType::new(type_row![NAT], type_row![NAT])
+                    .with_extension_delta(&just_a),
+            },
+            just_a.clone(),
+        ));
+
+        let root = hugr.root();
+
+        let exit = hugr.add_node_with_parent(
+            root,
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let entry = {
```

Comment:
One more create_with_io ? Or, better, why not make `make_block` return all three blocks